File: web-app/src/utils/ApiQuotaLedger.ts
/**
 * Tracks API request quotas over a rolling time window.
 *
 * The ledger increments with each request and determines
 * whether another call can be made based on the configured
 * limit.
 */
export class ApiQuotaLedger {
  private rollingCount = 0;
  private windowStart = Date.now();
  private limit: number;
  private windowMs: number;
  constructor(limit: number, windowMs = 24 * 60 * 60 * 1000) {
    this.limit = limit;
    this.windowMs = windowMs;
  }

  /**
   * Increase the count for the current window.
   *
   * Resets the window when the configured time span has
   * elapsed. This mutates the internal counters.
   */
  increment() {
    const now = Date.now();
    if (now - this.windowStart >= this.windowMs) {
      this.windowStart = now;
      this.rollingCount = 0;
    }
    this.rollingCount++;
  }

  /**
   * Check if another request can be made without exceeding the
   * quota.
   *
   * @returns `true` when under the limit or the time window has
   *   rolled over.
   */
  isSafe(): boolean {
    const now = Date.now();
    if (now - this.windowStart >= this.windowMs) {
      return true;
    }
    return this.rollingCount < this.limit;
  }
}

File: mobile-app/packages/services/lib/src/api_quota_ledger.dart
class ApiQuotaLedger {
  int _count = 0;
  DateTime _start = DateTime.now();
  final int limit;
  final Duration window;

  ApiQuotaLedger(this.limit, [this.window = const Duration(hours: 24)]);

  void increment() {
    final now = DateTime.now();
    if (now.difference(_start) >= window) {
      _start = now;
      _count = 0;
    }
    _count++;
  }

  bool isSafe() {
    final now = DateTime.now();
    if (now.difference(_start) >= window) {
      return true;
    }
    return _count < limit;
  }
}
